src/tactic/simps.lean

```diff
@@ -17,110 +17,318 @@ structures, projections, simp, simplifier, generates declarations
 
 open tactic expr
 
+setup_tactic_parser
+reserve notation `specify_simps_projections`
 declare_trace simps.verbose
 
+/--
+The `@[_simps_str]` attribute specifies the preferred projections of the given structure,
+used by the `@[simps]` attribute.
+- This will usually be tagged by the `@[simps]` tactic.
+- You can also generate this with the command `initialize_simps_projections`.
+- To change the default value, see Note [custom simps projection].
+- You are strongly discouraged to add this attribute manually.
+- The first argument is the list of names of the universe variables used in the structure
+- The second argument is the expressions that correspond to the projections of the structure
+  (these can contain the universe parameters specified in the first argument).
+-/
+@[user_attribute] meta def simps_str_attr : user_attribute unit (list name × list expr) :=
+{ name := `_simps_str,
+  descr := "An attribute specifying the projection of the given structure.",
+  parser := do e ← texpr, eval_pexpr _ e }
+
+/--
+  The `@[notation_class]` attribute specifies that this is a notation class,
+  and this notation should be used instead of projections by @[simps].
+  * The first argument `tt` for notation classes and `ff` for classes applied to the structure,
+    like `has_coe_to_sort` and `has_coe_to_fun`
+  * The second argument is the name of the projection (by default it is the first projection
+    of the structure)
+-/
+@[user_attribute] meta def notation_class_attr : user_attribute unit (bool × option name) :=
+{ name := `notation_class,
+  descr := "An attribute specifying that this is a notation class. Used by @[simps].",
+  parser := prod.mk <$> (option.is_none <$> (tk "*")?) <*> ident? }
+
+attribute [notation_class] has_zero has_one has_add has_mul has_inv has_neg has_sub has_div has_dvd
+  has_mod has_le has_lt has_append has_andthen has_union has_inter has_sdiff has_equiv has_subset
+  has_ssubset has_emptyc has_insert has_singleton has_sep has_mem has_pow
+
+attribute [notation_class* coe_sort] has_coe_to_sort
+attribute [notation_class* coe_fn] has_coe_to_fun
+
+/--
+  Finds an instance of an implication `cond → tgt`.
+  Returns a pair of a local constant `e` of type `cond`, and an instance of `tgt` that can mention `e`.
+-/
+meta def mk_conditional_instance (cond tgt : expr) : tactic (expr × expr) := do
+f ← mk_meta_var cond,
+e ← assertv `c cond f, swap,
+reset_instance_cache,
+inst ← mk_instance tgt,
+return (e, inst)
+
+/-- Get the projections used by `simps` associated to a given structure. The second component is the
```

Comment:
This is a monster (and important) function. The inline comments and tracing help a lot to follow it, but could you expand the docs a bit? Maybe giving an example of input alongside the corresponding output with a brief description of why it produced that output?